consensus/src/dag/dag_state_sync.rs

```diff
@@ -0,0 +1,180 @@
+use super::{
+    dag_fetcher::{DagFetcher, TDagFetcher},
+    dag_store::Dag,
+    storage::DAGStorage,
+    types::{CertifiedNodeWithLedgerInfo, RemoteFetchRequest},
+    TDAGNetworkSender,
+};
+use crate::state_replication::StateComputer;
+use anyhow::anyhow;
+use aptos_infallible::RwLock;
+use aptos_logger::error;
+use aptos_time_service::TimeService;
+use aptos_types::{
+    epoch_change::EpochChangeProof, epoch_state::EpochState, ledger_info::LedgerInfoWithSignatures,
+};
+use itertools::Itertools;
+use std::sync::Arc;
+
+pub const DAG_WINDOW: u64 = 10;
+
+pub(super) struct StateSyncManager {
+    epoch_state: Arc<EpochState>,
+    network: Arc<dyn TDAGNetworkSender>,
+    time_service: TimeService,
+    state_computer: Arc<dyn StateComputer>,
+    storage: Arc<dyn DAGStorage>,
+    dag_store: Arc<RwLock<Dag>>,
+}
+
+impl StateSyncManager {
+    pub fn new(
+        epoch_state: Arc<EpochState>,
+        network: Arc<dyn TDAGNetworkSender>,
+        time_service: TimeService,
+        state_computer: Arc<dyn StateComputer>,
+        storage: Arc<dyn DAGStorage>,
+        dag_store: Arc<RwLock<Dag>>,
+    ) -> Self {
+        Self {
+            epoch_state,
+            network,
+            time_service,
+            state_computer,
+            storage,
+            dag_store,
+        }
+    }
+
+    pub async fn sync_to(
+        &self,
+        node: &CertifiedNodeWithLedgerInfo,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        self.sync_to_highest_commit_cert(node.ledger_info()).await;
+        self.try_sync_to_highest_ordered_anchor(node).await
+    }
+
+    /// Fast forward in the decoupled-execution pipeline if the block exists there
+    pub async fn sync_to_highest_commit_cert(&self, ledger_info: &LedgerInfoWithSignatures) {
+        let dag_reader = self.dag_store.read();
+
+        // if the anchor exists between ledger info round and highest ordered round
+        // Note: ledger info round <= highest ordered round
+        if dag_reader.highest_committed_round().unwrap_or_default()
+            < ledger_info.commit_info().round()
+            && dag_reader.exists_by_round_digest(
```

Comment:
do we need this existence check? if the round is between commit and order, it should always exist if we don't violate safety?